### ISSUE TITLE
proposal: add lint or collection of lints to detect when using interfaces in an impure way.

### ISSUE THREAD

**Issue by mmcdon20** [dart-lang/linter#4963]
Dart allows you to to use interfaces in an impure way (especially within the same library it is defined). It would be nice if there were lints that could detect and warn against any of the following:

1. class fields in an interface
2. constructors in an interface
3. method bodies in an interface
4. a class extending an interface (unless that class is also an interface)
5. an interface class that is not also abstract
6. creating an instance of an interface

## Kind

Style advice

## Bad Examples

```dart
interface class Vehicle { // 5. an interface class that is not also abstract
  final String make; // 1. class fields in an interface
  final String model; // 1. class fields in an interface
  Vehicle(this.make, this.model); // 2. constructors in an interface
  
  void moveForward() { // 3. method bodies in an interface
    print('$make $model is moving forward!');
  }
}

class HondaCivic extends Vehicle { // 4. a class extending an interface 
  HondaCivic(): super('Honda', 'Civic');
}

void main() {
  var vehicle = Vehicle('Toyota', 'Camry'); // 6. creating an instance of an interface
  print(vehicle.make);
}
```

## Good Examples 

```dart
abstract interface class Vehicle {
  String get make;
  String get model;
  void moveForward();
}

class HondaCivic implements Vehicle {
  @override
  final String make = 'Honda';
  @override
  final String model = 'Civic';
  @override
  void moveForward() {
    print('$make $model is moving forward!');
  }
}
```

## Discussion

See also https://github.com/dart-lang/language/issues/3736 and https://github.com/dart-lang/sdk/issues/55641.

**Comment by lrhn** [comment]
it's not a lint I'd want in the recommended set. That is, it's a style lint that one can opt in to, but not an official style recommendation. 

About the individual points of warning: 

> - class fields in an interface

Should be "concrete instance field declaration". A declaration of
```dart
  abstract final int foo;
```
It's an instance variable declaration, but it doesn't introduce any implementation. 

> - constructors in an interface

Should probably only be _generative_ constructors. Having an interface with factory constructors creating instances of classes implementing the interface, is a common pattern.

> - method bodies in an interface

Can be combined with first item into:
"Non-abstract instance member declarations".

> - a class extending an interface (unless that class is also an interface)

Can only happen inside the same library.
I'd remove the exception. Other libraries cannot extend an `interface` declared class, which means that extending is not considered using it as an interface. This lint is about enforcing that the interface is only used as an interface.
So no exception. 


> - an interface class that is not also abstract

Should have no effect if there is no generative constructor, and required if there is any abstract instance member (what must be abstract, so any instance member at all.)
But does prevent invoking a default constructor.
So sure.

> - creating an instance of an interface

Redundant with not allowing generative constructors and being abstract.

Might also want:
- must not extend another class. (No `extends` or `with` clause.)

Otherwise the interface can inherit concrete instance members.

**Comment by mmcdon20** [comment]
> it's not a lint I'd want in the recommended set. That is, it's a style lint that one can opt in to, but not an official style recommendation.
> 
> About the individual points of warning:
> 
> > * class fields in an interface
> 
> Should be "concrete instance field declaration". A declaration of
> 
> ```dart
>   abstract final int foo;
> ```
> 
> It's an instance variable declaration, but it doesn't introduce any implementation.
> 

Good, point, `abstract` fields should be allowed.

> > * constructors in an interface
> 
> Should probably only be _generative_ constructors. Having an interface with factory constructors creating instances of classes implementing the interface, is a common pattern.
> 

Also good point, `factory` constructors and also static members would be okay.

> > * method bodies in an interface
> 
> Can be combined with first item into: "Non-abstract instance member declarations".
> 

Sure why not.

> > * a class extending an interface (unless that class is also an interface)
> 
> Can only happen inside the same library. I'd remove the exception. Other libraries cannot extend an `interface` declared class, which means that extending is not considered using it as an interface. This lint is about enforcing that the interface is only used as an interface. So no exception.
> 

This one I'm not so sure about. In Java and C# it is perfectly legal for an interface to extend another interface, I don't think that makes the use of interface impure. This seems like it would be a useful enough feature, although as you point out it is already not allowed outside of the same library which would make the rule seem inconsistent.

> > * an interface class that is not also abstract
> 
> Should have no effect if there is no generative constructor, and required if there is any abstract instance member (what must be abstract, so any instance member at all.) But does prevent invoking a default constructor. So sure.
> 

Okay

> > * creating an instance of an interface
> 
> Redundant with not allowing generative constructors and being abstract.
> 

Agreed

> Might also want:
> 
> * must not extend another class. (No `extends` or `with` clause.)
> 
> Otherwise the interface can inherit concrete instance members.

Touched on extending above, dart is unusual in that classes by default are also interfaces, and in other languages it is okay for an interface to extend an interface, but extending a class is weird, so maybe. 

No `with` makes sense.

**Comment by mmcdon20** [comment]
I have updated the proposal to account for @lrhn's feedback.
